lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -89,10 +89,52 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
   return valueAttr();
 }
 
+/// Check if the results of \p op are of integer type and with known bitwidth.
+/// Can be used to determine if any fold is legal.
+bool hasKnownWidths(Operation *op) {
+  for (auto resTy : op->getResultTypes()) {
+    auto ty = resTy.cast<FIRRTLType>();
+    if (!ty.isa<IntType>() || ty.getBitWidthOrSentinel() <= 0)
+      return false;
+  }
+  return true;
+}
+
+// Width extension of a constant int depends only on the sign of the operand.
+APInt extendConstantToWidth(APInt old, bool isSigned, unsigned newWidth) {
+  if (isSigned)
+    return old.sext(newWidth);
+  return old.zext(newWidth);
+}
+
 //===----------------------------------------------------------------------===//
 // Binary Operators
 //===----------------------------------------------------------------------===//
 
+OpFoldResult AddPrimOp::fold(ArrayRef<Attribute> operands) {
+  /// Any folding here requires a bitwidth extension.
+
+  /// If both operands are constant, and the result is integer with known
+  /// widths, then perform constant folding. Cannot use constFoldBinaryOp, since
+  /// the width of the constant is different from operands.
+  if (operands[0] && operands[1] && hasKnownWidths(*this)) {
+    auto op1 = operands[0].cast<IntegerAttr>();
+    auto op2 = operands[1].cast<IntegerAttr>();
+    auto numBits = getType().cast<FIRRTLType>().getBitWidthOrSentinel();
+    APInt sumAPInt =
+        extendConstantToWidth(
+            op1.getValue(),
+            getOperand(0).getType().cast<FIRRTLType>().isSignedInteger(),
```

Comment:
Could it be that this `isSignedInteger()` always returns `false`, regardless of the type? That would explain the arithmetic mismatches you see further down in your test cases, because this would cause operands to be zero-extended all the time.

Reply:
Thanks, that was the issue. I was wrongly using ``llvm::Type::isSignedInteger()``, updated it to ``FIRRTLType::IntType::isSigned()``